### ISSUE TITLE
API: document/col-name/_key and cursor return different floats

### ISSUE THREAD

**Issue by baslr** [arangodb/arangodb#1230]
Hi,

the api endpoint to retrieve a document based on its key returns incorrect floats.

```coffeescript
# v2.4.2

db.query.exec 'FOR t IN test FILTER t.test RETURN [t.test, t._key]'
{
  "result": [
    [
      4e-262,
      "2223512501955"
    ]
  ],
  "hasMore": false,
  "extra": {
    "stats": {
      "writesExecuted": 0,
      "writesIgnored": 0,
      "scannedFull": 4,
      "scannedIndex": 0
    },
    "warnings": []
  },
  "error": false,
  "code": 201
}


db.document.get 'test/2223512501955'
{
  "test": 3.9999999999999967e-262, # actually its 4e-262
  "_id": "test/2223512501955",
  "_rev": "2223518400195",
  "_key": "2223512501955",
  "code": 200
}
```

In the Web Interface:

/_db/_system/_api/cursor (collection overview)

![bildschirmfoto 2015-01-20 um 14 32 01](https://cloud.githubusercontent.com/assets/1494154/5818962/b6c0663a-a0b1-11e4-8ef8-5da5178bc368.png)
response
{"result":[{"_id":"test/2223512501955","_key":"2223512501955","_rev":"2223518400195","test":4e-262}],"hasMore":false,"extra":{"stats":{"writesExecuted":0,"writesIgnored":0,"scannedFull":4,"scannedIndex":0,"fullCount":4},"warnings":[]},"error":false,"code":201}

/_db/_system/_api/document/test/2223512501955?_=1422883365316 (single document)
![bildschirmfoto 2015-01-20 um 14 32 13](https://cloud.githubusercontent.com/assets/1494154/5818966/c30511c0-a0b1-11e4-9c9b-6b226dc96e8e.png)
response:
{"test":0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000039999999999999968,"_id":"test/2223512501955","_rev":"2223518400195","_key":"2223512501955"}

**Comment by jsteemann** [comment]
These are all rounding issues.

ArangoDB uses IEEE754 floating point numbers internally, and these cannot exactly represent each possible value. `0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000039999999999999968` and `3.9999999999999967e-262` are both approximations of `4e-262`, the first one in decimal format and the second in scientific notation. But it's approximately the same values. 
The confusion probably because of the two different string representations of the same numeric value.

Given that IEEE754 floating point values are in general prone to precision loss and rounding errors, all that can be done sensibly about this is to pass the numbers through the same to-string conversion and rounding mechanism in all places in our code. Then there is also V8, which uses its own IEEE754-to-string implementation, and this one we will not be able to adjust.

In general, if you require your numeric values to have arbitrary precision, it won't be safe to use IEEE754 binary floating points and rely on them, neither in ArangoDB, nor in other languages (JavaScript, PHP, etc. use IEEE754 floats, too).

For example, here's is my Firefox's view of 0.1 + 0.2, and if their sum is equivalent to 0.3:
```js
0.1 + 0.2
0.30000000000000004

0.1 + 0.2 === 0.3
false
```

V8 will come to the same conclusion.

**Comment by baslr** [comment]
`to pass the numbers through the same to-string conversion and rounding mechanism in all places in our code`
this would be a very good step. Is it possible to use the converter from the cursor call? Because this converts back to `4e-262` instead of `0.0...`

**Comment by 13abylon** [comment]
Duplicate of #1185

**Comment by jsteemann** [comment]
The *it returns different floats* issue should be fixed in 2.5.
